src/main/scala/tech/cryptonomic/conseil/Harpoon.scala

```diff
@@ -0,0 +1,142 @@
+import cats.effect.{ExitCode, IO, IOApp}
+import tech.cryptonomic.conseil.process.AccountsTaggingProcess
+import tech.cryptonomic.conseil.process.AccountsTaggingProcess.BlockLevel
+import tech.cryptonomic.conseil.tezos.{Tables, TezosDatabaseOperations, TezosTypes}
+import tech.cryptonomic.conseil.tezos.TezosDatabaseOperations.AccountFlags
+import tech.cryptonomic.conseil.util.{DatabaseUtil, PureLogging}
+import tech.cryptonomic.conseil.config.HarpoonAppConfig
+import slick.jdbc.PostgresProfile.api._
+import scala.concurrent.ExecutionContext
+import tech.cryptonomic.conseil.util.IOUtil
+import slick.jdbc.ResultSetType
+import slick.jdbc.ResultSetConcurrency
+import cats.data.StateT
+import cats.mtl.implicits._
+import cats.~>
+
+/** Runs an independet process that scans existing operations
+  * recorded from the chainm to look for reveal or activation
+  * and thus mark the corresponding Account record
+  */
+object Harpoon extends IOApp with PureLogging {
+
+  /** A type alias to provide state management on top of IO.
+    * The state threaded through corresponds to the actual
+    * process input value (AccountsTaggingProcess.ProcessorInput i.e. Option[BlockLevel]),
+    * which allows to start processing from that point onward
+    */
+  type IOState[T] = StateT[IO, Option[BlockLevel], T]
+
+  /* reference to the database to use, we need write access to update accounts */
+  private lazy val dbHandle: Database = DatabaseUtil.lorreDb
+
+  /* This is the main entrypoint, the return type is wrapped in IO and provides an ExitCode for the process
+   * Described in cats-effect apis: https://typelevel.org/cats-effect/api/cats/effect/IOApp.html
+   */
+  override def run(args: List[String]): IO[ExitCode] = {
+
+    //we read custom conf, usually provided from the environment
+    val conf = HarpoonAppConfig
+
+    implicit val ctx = scala.concurrent.ExecutionContext.global
+    val timer = IO.timer(scala.concurrent.ExecutionContext.global)
+
+    //we create the processor stage, with the custom types needed
+    val tagging = AccountsTaggingProcess[IOState]
+
+    /* Implement any dependency needed to run the processing. */
+    implicit val serviceApis = injectServices(tagging)
+
+    /* now we can call the processing stage */
+    val processCycle = (processedTopLevel: tagging.ProcessorInput) =>
+      tagging.process
+        .run(processedTopLevel)
+        .flatTap {
+          case (level, flagged) =>
+            logger.pureLog[IO](
+              _.info(
+                "This Harpoon processing cycle marked {} accounts, up to level {}.",
+                flagged.fold("an unknown number of")(String.valueOf),
+                level.fold("unspecified")(String.valueOf)
+              )
+            )
+        }
+        .handleErrorWith(
+          err =>
+            logger.pureLog[IO](_.error("Harpoon processing failed to complete", err)) *>
+                IO(processedTopLevel, 0) //we reset to the previous state as to retry next round
```

Comment:
Avoid autotupling - `IO((processedTopLevel, 0))` should be used instead (if there are no side effects here, `IO.pure` can be used too)

Reply:
Cool, mistake

Reply:
I actually never considered that there's a difference between `IO.apply` and `IO.pure`, great insight